src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -93,10 +93,11 @@ public Solution GetCompileTimeSolution(Solution designTimeSolution)
                 var staleSolution = _lastCompileTimeSolution;
                 var compileTimeSolution = designTimeSolution;
 
+                using var _1 = PooledHashSet<ProjectId>.GetInstance(out var projectsWithDesignTimeDocuments);
```

Comment:
>PooledHashSet [](http://example.com/codeflow?start=31&length=13)

Instead of using a HashSet we can use ArrayBuilder and then in the inner document loop set a boolean flag that's reset in the outer project iterating loop, where the project can be added to the array builder.